### PR TITLE
Add species clustering

### PR DESCRIPTION
**Pull request by meomancer**
this fix #https://github.com/kartoza/LEDET_BIMS/issues/64

This API can return clustering data on fly, with clustering by size of icon of point.
What this API:
- clustering data by icon with parameter : `icon_pixel_x=30` & `icon_pixel_y=30`
- need zoom of map parameter : `zoom=8`
- filter data by search query : `query=matar` (optional, default is all data)
- filter data by module : `module=fish` (optional, default is location_site)
- filter data by bbox of map : `bbox=14.243774414062505,-34.84536693184098,34.6234130859375,-29.118574414910853` (optional, default will return all data)

________________________________________
### Example: zoom 7
http://localhost:63302/api/cluster-module/?icon_pixel_x=30&icon_pixel_y=30&module=fish&zoom=7&bbox=14.243774414062505,-34.84536693184098,34.6234130859375,-29.118574414910853

all is in cluster
selection_036

________________________________________
### Example: zoom 8
http://localhost:63302/api/cluster-module/?icon_pixel_x=30&icon_pixel_y=30&module=fish&zoom=8&bbox=14.243774414062505,-34.84536693184098,34.6234130859375,-29.118574414910853

There is in cluster
selection_037

and individual data
selection_038

________________________________________
TODO and Question:
- Fixing location sites elasticsearch
- [QUESTION] currently for the cluster, the position is the center of bbox of cluster. Should we change?
- @gubuntu due to big data, instead show per layer, how about we combine in 1 api? if user select fish and coral, for example, this api will return clustering (and data) for just both of these module.
- We need to enable location search on records, for map bbox filtering (currently, i was checking returned data from elastic search location)
- need to OR query on this : https://github.com/kartoza/django-bims/pull/36/files#diff-08260bd2eee44903496a79e9e6131395R258. But i still can't done it.

### CODE REVIEW

**Comment by meomancer** [comment]
- add `module=taxa`, will search all record
- using django search instead using elasticsearch


http://localhost:63302/api/cluster-module/?icon_pixel_x=30&icon_pixel_y=30&module=taxa&zoom=8&bbox=14.243774414062505,-34.84536693184098,34.6234130859375,-29.118574414910853

selection_039

__________________________________
with `query=moss`

http://localhost:63302/api/cluster-module/?icon_pixel_x=30&icon_pixel_y=30&module=taxa&zoom=8&bbox=14.243774414062505,-34.84536693184098,34.6234130859375,-29.118574414910853&query=moss

selection_040
______________________________________

with `query=bengalens`
http://localhost:63302/api/cluster-module/?icon_pixel_x=30&icon_pixel_y=30&module=taxa&zoom=8&bbox=14.243774414062505,-34.84536693184098,34.6234130859375,-29.118574414910853&query=bengalens
selection_041

When zoom in:
selection_042

**Comment by meomancer** [comment]
Update clustering by same taxa:
- receive pk id of record
- from that record, get taxa and clustering all record from that taxa

With zoom 18
http://localhost:63302/api/cluster-data/taxa/8/?icon_pixel_x=30&icon_pixel_y=30&zoom=18&bbox=14.243774414062505,-34.84536693184098,34.6234130859375,-29.118574414910853
selection_044

Zoom 8
selection_045


Taxa 1 : 
http://localhost:63302/api/cluster-data/taxa/1/?icon_pixel_x=30&icon_pixel_y=30&zoom=18&bbox=14.243774414062505,-34.84536693184098,34.6234130859375,-29.118574414910853
selection_046

File: bims/static/js/views/olmap.js [hunk 1]
```diff
@@ -181,11 +181,13 @@ define([
...
                 var zoomLevel = self.map.getView().getZoom();
-                if(zoomLevel !== self.previousZoom) {
+                if (zoomLevel !== self.previousZoom) {
                     self.previousZoom = zoomLevel;
                     self.refetchCollection();
                 }
-
+                var ext = self.map.getView().calculateExtent(self.map.getSize())
+                ext = ol.proj.transformExtent(ext, ol.proj.get('EPSG:3857'), ol.proj.get('EPSG:4326'));
+                console.log('http://localhost:63302/api/cluster-module/?icon_pixel_x=30&icon_pixel_y=30&module=fish' + '&zoom=' + self.map.getView().getZoom() + '&bbox=' + ext);
```
Comment (timlinux) [inline bims/static/js/views/olmap.js]: This will come out in production right?

**Comment by timlinux** [comment]
Looking good @meomancer 

Can you please create a test suite for this 

Where are you taking the taxon from for the clustering? BiologicalCollection or Taxon table?

**Comment by dimasciput** [comment]
@timlinux 

> Where are you taking the taxon from for the clustering? BiologicalCollection or Taxon table?

It will filters all the biologicalCollection based on the taxa id.

**Comment by meomancer** [comment]
hi @timlinux @dimasciput 
basically, it receive pk on BiologicalCollection, after that get all BiologicalCollection that same taxon id with the received collection, and it is return BiologicalCollection

**Comment by meomancer** [comment]
Did stress test with big zoom. The peformance is good.
Collection records is 100.

|request|time|
|-------|----|
|/api/cluster/collection/taxon/2/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|28 ms|
|/api/cluster/collection/taxon/3/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|33 ms|
|/api/cluster/collection/taxon/4/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|34 ms|
|/api/cluster/collection/taxon/5/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|100 ms|
|/api/cluster/collection/taxon/6/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|28 ms|
|/api/cluster/collection/taxon/7/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|34 ms|
|/api/cluster/collection/taxon/8/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|70 ms|
|/api/cluster/collection/taxon/9/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|31 ms|
|/api/cluster/collection/taxon/10/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|79 ms|
|/api/cluster/collection/taxon/11/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|28 ms|
|/api/cluster/collection/taxon/12/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|27 ms|
|/api/cluster/collection/taxon/13/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|336 ms|
|/api/cluster/collection/taxon/14/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|34 ms|
|/api/cluster/collection/taxon/15/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|31 ms|
|/api/cluster/collection/taxon/16/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|41 ms|
|/api/cluster/collection/taxon/17/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|38 ms|
|/api/cluster/collection/taxon/18/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|32 ms|
|/api/cluster/collection/taxon/19/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|37 ms|
|/api/cluster/collection/taxon/20/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|48 ms|
|/api/cluster/collection/taxon/21/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|41 ms|
|/api/cluster/collection/taxon/22/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|34 ms|
|/api/cluster/collection/taxon/23/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|107 ms|
|/api/cluster/collection/taxon/24/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|529 ms|
|/api/cluster/collection/taxon/25/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|55 ms|
|/api/cluster/collection/taxon/26/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|37 ms|
|/api/cluster/collection/taxon/27/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|26 ms|
|/api/cluster/collection/taxon/28/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|28 ms|
|/api/cluster/collection/taxon/29/?icon_pixel_x=30&icon_pixel_y=30&zoom=7&bbox=11.154669085937511,-36.36119970817674,31.534307757812506,-29.42243402059367|106 ms|

**Comment by dimasciput** [comment]
LGTM

**Pull request merged**